src/modules/languages/rust.nix

```diff
@@ -18,62 +15,94 @@ in
   options.languages.rust = {
     enable = lib.mkEnableOption "tools for Rust development";
 
-    packages = lib.mkOption {
-      type = lib.types.submodule ({
-        options = {
-          rust-src = lib.mkOption {
-            type = lib.types.either lib.types.package lib.types.str;
-            default = pkgs.rustPlatform.rustLibSrc;
-            defaultText = lib.literalExpression "pkgs.rustPlatform.rustLibSrc";
-            description = "rust-src package";
-          };
-        }
-        // genAttrs tools (name: lib.mkOption {
-          type = lib.types.package;
-          default = pkgs.${name};
-          defaultText = lib.literalExpression "pkgs.${name}";
-          description = "${name} package";
-        });
-      });
-      defaultText = lib.literalExpression "pkgs";
-      default = { };
-      description = "Attribute set of packages including rustc and Cargo.";
+    package = lib.mkOption {
+      type = lib.types.package;
+      defaultText = lib.literalExpression "nixpkgs";
+      default = pkgs.symlinkJoin {
+        name = "nixpkgs-rust";
+        paths = with pkgs; [
+          rustc
+          cargo
+          rustfmt
+          clippy
+          rust-analyzer
+        ];
+      };
+      description = "Rust package including rustc and Cargo.";
+    };
+
+    components = lib.mkOption {
+      type = lib.types.listOf lib.types.str;
+      default = [ "rustc" "cargo" "clippy" "rustfmt" "rust-analyzer" ];
+      defaultText = lib.literalExpression ''[ "rustc" "cargo" "clippy" "rustfmt" "rust-analyzer" ]'';
+      description = ''
+        List of [Rustup components](https://rust-lang.github.io/rustup/concepts/components.html)
+        to install. Defaults to those available in `nixpkgs`.
+      '';
+    };
+
+    rust-src = lib.mkOption {
+      type = lib.types.path;
+      default = pkgs.rustPlatform.rustLibSrc;
+      defaultText = "`pkgs.rustPlatform.rustLibSrc` or `toolchain.rust-src`, depending on if a fenix toolchain is set.";
+      description = ''
+        The path to the rust-src Rustup component. Note that this is necessary for some tools
+        like rust-analyzer to work. See [Rustup docs](https://rust-lang.github.io/rustup/concepts/components.html)
+        for more information.
+      '';
+    };
+
+    toolchain = lib.mkOption {
+      # TODO: better type
+      type = lib.types.nullOr (lib.types.attrsOf lib.types.anything);
+      default = null;
+      defaultText = lib.literalExpression "fenix.packages.stable";
+      description = "The [fenix toolchain](https://github.com/nix-community/fenix#toolchain) to use.";
     };
 
     version = lib.mkOption {
```

Comment:
Nit, but I wonder if we should really rename this to `channel`. You can't really specify a version with `fenix` because it's lacking the rust version manifests. `rust-overlay` has these though if we wanted to support that.

Reply:
tbh i haven't looked at `rust-overlay` before, but i think it can provide a better api for devenv:

```nix
# use stable, beta, or nightly toolchain
channel = "stable";

# specify profile
profile = "default";

# you can still use `package` for more flexibility in `fenix` or `rust-overlay`
package = pkgs.fenix.stable.toolchain;
```

Reply:
i did originally have this renamed to `channel` but the names aren't the same as rustup channels, which is why i decided to keep it as `version`. i agree, it's not really fitting, but it's more accurate than `channel`. after further thinking, i do think the correct solution is to switch to `rust-overlay` which is a bit more accurate in its naming